File: backend/app/routers/attendance.py
from datetime import date, datetime
from typing import Optional
from fastapi import APIRouter, Body, Depends, Query, status
from sqlalchemy.orm import Session

from app.core.dependencies import get_db, require_role
from app.schemas.attendance import (
    AttendanceListResponse,
    AttendanceResponse,
    CheckInRequest,
    CheckOutRequest,
    DailyAttendanceSummaryResponse,
)
from app.services.attendance_service import attendance_service

# Attendance endpoints require either ADMIN or HR role privileges
router = APIRouter(
    prefix="/attendance",
    tags=["Attendance"],
    dependencies=[Depends(require_role("ADMIN", "HR"))],
)


@router.post(
    "",
    response_model=AttendanceResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Check-in employee attendance",
    description="Mark attendance for an active employee. Checks for existing check-in on the same date and enforces timezone-aware timestamps in Asia/Kolkata.",
)
def check_in(
    payload: CheckInRequest,
    db: Session = Depends(get_db),
):
    """Mark attendance check-in for an employee."""
    return attendance_service.check_in(db, payload)


# Route aliases for check-in
@router.post(
    "/check-in",
    response_model=AttendanceResponse,
    status_code=status.HTTP_201_CREATED,
    include_in_schema=False,
)
@router.post(
    "/checkin",
    response_model=AttendanceResponse,
    status_code=status.HTTP_201_CREATED,
    include_in_schema=False,
)
def check_in_alias(
    payload: CheckInRequest,
    db: Session = Depends(get_db),
):
    """Alias route for employee check-in."""
    return attendance_service.check_in(db, payload)


@router.put(
    "/{attendance_id}/checkout",
    response_model=AttendanceResponse,
    status_code=status.HTTP_200_OK,
    summary="Check-out employee attendance",
    description="Record check-out for an existing attendance record using the attendance primary key ID.",
)
def check_out(
    attendance_id: int,
    payload: Optional[CheckOutRequest] = None,
    db: Session = Depends(get_db),
):
    """Record attendance check-out using attendance record ID."""
    check_out_val = payload.check_out if payload else None
    return attendance_service.check_out_by_id(db, attendance_id, check_out_val)


@router.get(
    "/daily-status",
    response_model=DailyAttendanceSummaryResponse,
    status_code=status.HTTP_200_OK,
    summary="Get daily attendance status and summary",
    description="Retrieve daily summary with computed PRESENT and ABSENT counts and employee list for a given date in Asia/Kolkata.",
)
def get_daily_status(
    attendance_date: Optional[date] = Query(None, description="Date (YYYY-MM-DD), defaults to today in Asia/Kolkata"),
    department_id: Optional[int] = Query(None, description="Filter by department ID"),
    page: int = Query(1, ge=1, description="Page number (1-indexed)"),
    page_size: int = Query(10, ge=1, le=100, description="Items per page"),
    db: Session = Depends(get_db),
):
    """Retrieve daily attendance status breakdown for active employees."""
    return attendance_service.get_daily_status(
        db=db,
        attendance_date=attendance_date,
        department_id=department_id,
        page=page,
        page_size=page_size,
    )


@router.get(
    "",
    response_model=AttendanceListResponse,
    status_code=status.HTTP_200_OK,
    summary="List, filter, and paginate attendance records",
    description="Retrieve paginated attendance records with optional filtering by date, date range, employee, and department, with whitelisted deterministic sorting.",
)
def list_attendance(
    page: int = Query(1, ge=1, description="Page number (1-indexed)"),
    page_size: int = Query(10, ge=1, le=100, description="Number of items per page (max 100)"),
    attendance_date: Optional[date] = Query(None, description="Exact attendance date (YYYY-MM-DD)"),
    start_date: Optional[date] = Query(None, description="Start date for range filtering (inclusive)"),
    end_date: Optional[date] = Query(None, description="End date for range filtering (inclusive)"),
    employee_id: Optional[str] = Query(None, description="Filter by human-facing employee ID"),
    department_id: Optional[int] = Query(None, description="Filter by department ID"),
    include_absent: bool = Query(
        False,
        description="When attendance_date is set, includes active employees without records computed as ABSENT",
    ),
    sort_by: str = Query(
        "attendance_date",
        description="Sort column: 'attendance_date', 'check_in', 'check_out', 'created_at', 'employee_id'",
    ),
    sort_order: str = Query(
        "desc",
        description="Sort direction: 'asc', 'desc'",
    ),
    db: Session = Depends(get_db),
):
    """List attendance records according to query filters and pagination."""
    return attendance_service.list_attendance(
        db=db,
        page=page,
        page_size=page_size,
        attendance_date=attendance_date,
        start_date=start_date,
        end_date=end_date,
        employee_id=employee_id,
        department_id=department_id,
        include_absent=include_absent,
        sort_by=sort_by,
        sort_order=sort_order,
    )

File: frontend/src/api/attendance.js
import apiClient from './client';

/**
 * Returns today's date formatted as YYYY-MM-DD in Asia/Kolkata timezone.
 * @returns {string}
 */
export function getTodayKolkataDateString() {
  const formatter = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'Asia/Kolkata',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  });
  return formatter.format(new Date());
}

/**
 * Formats a timezone-aware ISO string to readable local time in Asia/Kolkata (e.g. "09:30 AM").
 * @param {string|null} dateString
 * @returns {string}
 */
export function formatTimeInKolkata(dateString) {
  if (!dateString) return '—';
  try {
    const d = new Date(dateString);
    return d.toLocaleTimeString([], {
      hour: '2-digit',
      minute: '2-digit',
      hour12: true,
      timeZone: 'Asia/Kolkata',
    });
  } catch {
    return '—';
  }
}

/**
 * Formats a date string to readable format in Asia/Kolkata (e.g. "Sep 5, 2026").
 * @param {string|null} dateString
 * @returns {string}
 */
export function formatDateInKolkata(dateString) {
  if (!dateString) return '—';
  try {
    const d = new Date(dateString.includes('T') ? dateString : `${dateString}T00:00:00`);
    return d.toLocaleDateString([], {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      timeZone: 'Asia/Kolkata',
    });
  } catch {
    return dateString;
  }
}

/**
 * Fetches paginated, filtered attendance records.
 * @param {Object} params
 * @param {number} [params.page=1]
 * @param {number} [params.pageSize=10]
 * @param {string} [params.attendanceDate='']
 * @param {string} [params.startDate='']
 * @param {string} [params.endDate='']
 * @param {string} [params.employeeId='']
 * @param {number|string} [params.departmentId='']
 * @param {boolean} [params.includeAbsent=false]
 * @param {string} [params.sortBy='attendance_date']
 * @param {string} [params.sortOrder='desc']
 * @returns {Promise<{ items: Array, total: number, page: number, page_size: number, total_pages: number }>}
 */
export async function getAttendanceApi({
  page = 1,
  pageSize = 10,
  attendanceDate = '',
  startDate = '',
  endDate = '',
  employeeId = '',
  departmentId = '',
  includeAbsent = false,
  sortBy = 'attendance_date',
  sortOrder = 'desc',
} = {}) {
  const queryParams = {
    page,
    page_size: pageSize,
    sort_by: sortBy,
    sort_order: sortOrder,
  };

  if (attendanceDate) queryParams.attendance_date = attendanceDate;
  if (startDate) queryParams.start_date = startDate;
  if (endDate) queryParams.end_date = endDate;
  if (employeeId && employeeId.trim()) queryParams.employee_id = employeeId.trim();
  if (departmentId && Number(departmentId) > 0) queryParams.department_id = Number(departmentId);
  if (includeAbsent) queryParams.include_absent = true;

  const response = await apiClient.get('/attendance', { params: queryParams });
  return response.data;
}

/**
 * Marks attendance check-in for an active employee.
 * @param {Object} payload
 * @param {string} payload.employee_id
 * @param {string} [payload.attendance_date]
 * @param {string} [payload.check_in]
 * @returns {Promise<Object>}
 */
export async function checkInApi({ employee_id, attendance_date, check_in }) {
  const payload = {
    employee_id: employee_id.trim(),
  };

  if (attendance_date) payload.attendance_date = attendance_date;
  if (check_in) payload.check_in = check_in;

  const response = await apiClient.post('/attendance', payload);
  return response.data;
}

/**
 * Records check-out for an existing attendance record by primary key ID.
 * @param {number} attendanceId
 * @param {Object} [payload={}]
 * @param {string} [payload.check_out]
 * @returns {Promise<Object>}
 */
export async function checkOutApi(attendanceId, { check_out } = {}) {
  const payload = {};
  if (check_out) payload.check_out = check_out;

  const response = await apiClient.put(`/attendance/${attendanceId}/checkout`, payload);
  return response.data;
}

/**
 * Retrieves daily attendance summary for active employees with computed PRESENT and ABSENT counts.
 * @param {Object} [params={}]
 * @param {string} [params.attendanceDate='']
 * @param {number|string} [params.departmentId='']
 * @param {number} [params.page=1]
 * @param {number} [params.pageSize=10]
 * @returns {Promise<{ attendance_date: string, total_active_employees: number, present_count: number, absent_count: number, items: Array, total: number, page: number, page_size: number, total_pages: number }>}
 */
export async function getDailyStatusApi({
  attendanceDate = '',
  departmentId = '',
  page = 1,
  pageSize = 10,
} = {}) {
  const queryParams = {
    page,
    page_size: pageSize,
  };

  if (attendanceDate) queryParams.attendance_date = attendanceDate;
  if (departmentId && Number(departmentId) > 0) queryParams.department_id = Number(departmentId);

  const response = await apiClient.get('/attendance/daily-status', { params: queryParams });
  return response.data;
}

/**
 * Retrieves chronological attendance history for an employee.
 * @param {string} employeeIdentifier
 * @param {Object} [params={}]
 * @param {number} [params.page=1]
 * @param {number} [params.pageSize=10]
 * @param {string} [params.startDate='']
 * @param {string} [params.endDate='']
 * @param {string} [params.sortOrder='desc']
 * @returns {Promise<{ items: Array, total: number, page: number, page_size: number, total_pages: number }>}
 */
export async function getEmployeeAttendanceHistoryApi(
  employeeIdentifier,
  { page = 1, pageSize = 10, startDate = '', endDate = '', sortOrder = 'desc' } = {}
) {
  const queryParams = {
    page,
    page_size: pageSize,
    sort_order: sortOrder,
  };

  if (startDate) queryParams.start_date = startDate;
  if (endDate) queryParams.end_date = endDate;

  const response = await apiClient.get(
    `/employees/${encodeURIComponent(employeeIdentifier.trim())}/attendance`,
    { params: queryParams }
  );
  return response.data;
}

/**
 * Calculates human-readable shift duration between check-in and check-out timestamps.
 * @param {string|null} checkInStr
 * @param {string|null} checkOutStr
 * @returns {string}
 */
export function calculateShiftDuration(checkInStr, checkOutStr) {
  if (!checkInStr || !checkOutStr) return '—';
  try {
    const start = new Date(checkInStr).getTime();
    const end = new Date(checkOutStr).getTime();
    if (isNaN(start) || isNaN(end) || end < start) return '—';
    const totalMinutes = Math.floor((end - start) / 60000);
    const hours = Math.floor(totalMinutes / 60);
    const minutes = totalMinutes % 60;
    if (hours === 0) return `${minutes}m`;
    return `${hours}h ${minutes}m`;
  } catch {
    return '—';
  }
}
